tests/storage/test_filesystem.py

```diff
@@ -0,0 +1,139 @@
+import unittest
+import tempfile
+import os
+import codecs
+from contextlib import contextmanager
+
+from storage import FileSystem
+
+
+@contextmanager
+def create_file_system(groups_list=None, write_groups_file=True):
+    sdir = tempfile.mkdtemp()
+    fs = FileSystem(sdir, groups_list)
+    if write_groups_file and groups_list:
+        with open(os.path.sep.join([fs.storage_dir, "groups.txt"]), "w") as f:
+            f.writelines(groups_list)
+
+    yield fs
+
+
+class TestFileSystem(unittest.TestCase):
+    def test_constructor_without_groups_list(self):
+        with create_file_system() as fs:
+            self.assertEqual([], list(os.scandir(fs.storage_dir)))
+
+    def test_constructor_with_groups_list(self):
+        with create_file_system(("1",), write_groups_file=False) as fs:
+            dirs_to_be_created = [
+                os.path.sep.join([fs.storage_dir,  "1"]),
+                os.path.sep.join([fs.storage_dir, "1", "confirm_codes"]),
+                os.path.sep.join([fs.storage_dir, "1", "confirmed"]),
+            ]
+
+            for d in dirs_to_be_created:
+                self.assertTrue(
+                    os.path.exists(d),
+                    f"no subdirectory created: {d}",
+                )
+
+            self.assertFalse(os.path.exists(os.path.sep.join([fs.storage_dir, "groups.txt"])))
+
+    def test_constructor_with_groups_list_txt(self):
+        with create_file_system(("1",), write_groups_file=True) as fs:

```

Comment:
Второй параметр можно было бы и не писать
```suggestion
        with create_file_system(("1",)) as fs:

```